### PR TITLE
sys/riotboot/flashwrite: add functions to invalid slots

### PR DESCRIPTION
**Pull request by firas-hamdi**
- riotboot/flashwrite: Add APIs to invalidate a specific header and invalidate the newest firmware version

### CODE REVIEW

File: sys/riotboot/flashwrite.c [hunk 1]
```diff
@@ -150,6 +151,34 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
...
+    flashpage_write_raw((void *) riotboot_slot_get_hdr(slot), (const void *) data_flash, 4);
+
+    return 0;
+}
+
+int riotboot_flashwrite_update_revert(void)
+{
+    int _slot_to_revert = -1;
+    _slot_to_revert = (riotboot_slot_get_hdr(riotboot_slot_other())->version
+            > riotboot_slot_get_hdr(riotboot_slot_current())->version) ? riotboot_slot_other() : riotboot_slot_current();
```
Comment (fjmolinas) [inline sys/riotboot/flashwrite.c:178]: Is checking the version enough? Should it check if the version is valid?

File: sys/include/riotboot/flashwrite.h [hunk 2]
```diff
@@ -128,6 +128,11 @@ typedef struct {
  */
 #define RIOTBOOT_FLASHWRITE_SKIPLEN     sizeof(RIOTBOOT_MAGIC)
 
+/**
+ * @brief Magic number used to invalidate a slot
+ */
+#define INVALIDATE_HDR                  0
```
Comment (fjmolinas) [inline sys/include/riotboot/flashwrite.h:134]: This would depend on the flash, in some cases `0x00` is the erase state in others `0xFF`, and therefore this could be a `NOP` depending on the operation.

File: sys/riotboot/flashwrite.c [hunk 3]
```diff
@@ -150,6 +151,34 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
...
 }
 
+int riotboot_flashwrite_invalidate(int slot)
+{
+    extern const unsigned riotboot_slot_numof;
+
+    if (riotboot_slot_numof == 1){
+        LOG_WARNING(LOG_PREFIX "Only one slot configured\n");
+        return -1;
+    }
```
Comment (fjmolinas) [inline sys/riotboot/flashwrite.c:161]: This behavior does not match the function documentation, @bergzand do you think it should allow invalidating in this case?

File: sys/riotboot/flashwrite.c [hunk 4]
```diff
@@ -150,6 +151,34 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
...
+    if ((riotboot_slot_validate(riotboot_slot_other()) != 0) || (riotboot_slot_validate(riotboot_slot_current()) != 0)){
+        LOG_INFO(LOG_PREFIX "There will be no valid images to run after reboot\n");
+        return -2;
+    }
```
Comment (fjmolinas) [inline sys/riotboot/flashwrite.c:165]: Same here, if it returns then it will not actually invalidate.

File: sys/riotboot/flashwrite.c [hunk 5]
```diff
@@ -150,6 +151,34 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
...
+
+    uint8_t data_flash[4] = { INVALIDATE_HDR };
+
+    flashpage_write_raw((void *) riotboot_slot_get_hdr(slot), (const void *) data_flash, 4);
```
Comment (fjmolinas) [inline sys/riotboot/flashwrite.c:169]:
```suggestion
    flashpage_write_raw((void *) riotboot_slot_get_hdr(slot), (const void *) data_flash, sizeof(RIOTBOOT_MAGIC));
```

File: sys/include/riotboot/flashwrite.h (hunk 2 above)
Comment (fjmolinas) [reply sys/include/riotboot/flashwrite.h:134]: Maybe we should add a `FLASH_ERASE_STATE` value, and use that as the write value. What do you think @bergzand should we add something like that?

File: sys/suit/transport/coap.c [hunk 6]
```diff
@@ -97,6 +97,7 @@ static inline void _print_download_progress(suit_manifest_t *manifest,
         }
     }
 #endif
+    (void) image_size;
```
Comment (fjmolinas) [inline sys/suit/transport/coap.c:100]: This change should be in its own commit IMO

File: sys/riotboot/flashwrite.c [hunk 7]
```diff
@@ -162,6 +191,7 @@ int riotboot_flashwrite_finish_raw(riotboot_flashwrite_t *state,
 #if CONFIG_RIOTBOOT_FLASHWRITE_RAW
     memcpy(state->firstblock_buf, bytes, len);
     flashpage_write(slot_start, state->firstblock_buf, RIOTBOOT_FLASHPAGE_BUFFER_SIZE);
+    res = 0;
```
Comment (fjmolinas) [inline sys/riotboot/flashwrite.c:194]: This one should not be needed, I think it is already fixed upstream, should go away with a rebase.

**Review by fjmolinas — CHANGES_REQUESTED** [review]
Hi @firas-hamdi thanks for your contribution and welcome to RIOT!

Some questions regarding this change, what is the use case you have for this? Otherwise I don't have anything against the change. Some comments and questions below.

Also have you taken a look at out [Commit conventions](https://github.com/miri64/RIOT/blob/master/CONTRIBUTING.md#commit-conventions)? Would you min taking a look to reformat the commits? I think here you could have 2 commits:

- the one adding the new api functions
- the one fixing the unused `image_size`

**Comment by fjmolinas** [comment]
> Also have you taken a look at out Commit conventions? Would you min taking a look to reformat the commits? I think here you could have 2 commits:
> 
>     the one adding the new api functions
>     the one fixing the unused image_size

We would deal with this at the end of the review of course, please consider taking a look at the whole [Contributing Guidleines](https://github.com/miri64/RIOT/blob/master/CONTRIBUTING.md#contributing-to-riot) :)

File: sys/riotboot/flashwrite.c [hunk 8]
```diff
@@ -150,6 +151,34 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
...
+    if (riotboot_slot_numof == 1){
+        LOG_WARNING(LOG_PREFIX "Only one slot configured\n");
+        return -1;
+    }
+    if ((riotboot_slot_validate(riotboot_slot_other()) != 0) || (riotboot_slot_validate(riotboot_slot_current()) != 0)){
+        LOG_INFO(LOG_PREFIX "There will be no valid images to run after reboot\n");
+        return -2;
+    }
+
+    uint8_t data_flash[4] = { INVALIDATE_HDR };
```
Comment (benpicco) [inline sys/riotboot/flashwrite.c:167]: This will only set the first element to `INVALIDATE_HDR` and the others to `0`.
Granted, the signature will be invalidated if even a single character is changed, but what you were trying to achieve can be done with 

```suggestion
    uint8_t data_flash[RIOTBOOT_MAGIC];
    memset(data_flash, INVALIDATE_HDR, sizeof(RIOTBOOT_MAGIC));
```

File: sys/include/riotboot/flashwrite.h (hunk 2 above)
Comment (benpicco) [reply sys/include/riotboot/flashwrite.h:134]: we could also use `0xAA`, that should always clear half the bits of the signature.

File: sys/riotboot/flashwrite.c [hunk 9]
```diff
@@ -150,6 +151,34 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
     return 0;
 }
 
+int riotboot_flashwrite_invalidate(int slot)
+{
+    extern const unsigned riotboot_slot_numof;
```
Comment (benpicco) [inline sys/riotboot/flashwrite.c:156]: that's already defined in `riotboot/slot.h`

File: sys/riotboot/flashwrite.c (hunk 1 above)
Comment (firas-hamdi) [reply sys/riotboot/flashwrite.c:178]: No checking the version number is not enough, that's right. The version validation is done after selecting the slot to revert, using the **riotboot_flashwrite_invalidate** API

File: sys/riotboot/flashwrite.c (hunk 3 above)
Comment (firas-hamdi) [reply sys/riotboot/flashwrite.c:161]: This check is done to make sure the device still has at least one firmware image to be running after invalidation, so a reboot after calling the API won't make the device with no image to run. And thus the warning generated to not allow the invalidation in this case.

File: sys/riotboot/flashwrite.c (hunk 4 above)
Comment (firas-hamdi) [reply sys/riotboot/flashwrite.c:165]: Yes to avoid the "no image" case after reboot

**Comment by firas-hamdi** [comment]
Hi @fjmolinas thank you for the review.
The use case is **to revert an update** ( due to false new image, or the need to run an old image) by invalidating the higher firmware version, so the lower version will be running after reboot. 
Invalidating a specific header allows in general to eliminate a specific image from the "to boot" list

File: sys/riotboot/flashwrite.c (hunk 3 above)
Comment (benpicco) [reply sys/riotboot/flashwrite.c:161]: I think this is fine, but the API documentation should reflect that behavior.

**Comment by fjmolinas** [comment]
> Hi @fjmolinas thank you for the review.
> The use case is **to revert an update** ( due to false new image, or the need to run an old image) by invalidating the higher firmware version, so the lower version will be running after reboot.
> Invalidating a specific header allows in general to eliminate a specific image from the "to boot" list

Well the function name make it clear what they are intended to do, my question was more on a meta level, and more on the curiosity side. You are making the changes in `flashwrite` and `suit`, so it made me think you are using both. But if its the case and the image is 'false' then a suit update will not complete so the image would not really need invalidation. Running an old image would also go against 'suit' so I don't see how these functions would fit with 'siot'. So it makes me think you are using `flashwrite` stand alone, and here is where I'm curious :).

**Comment by firas-hamdi** [comment]
> Well the function name make it clear what they are intended to do, my question was more on a meta level, and more on the curiosity side. You are making the changes in `flashwrite` and `suit`, so it made me think you are using both. But if its the case and the image is 'false' then a suit update will not complete so the image would not really need invalidation. Running an old image would also go against `suit` so I don't see how these functions would fit with `suit`. So it makes me think you are using `flashwrite` stand alone, and here is where I'm curious :).
>
The only change made in `suit` module is to fix the bug, I will make it in its own pull request. 
I was meaning by the term **false** I used after completing the suit update: example a bug discovered in the new image and a fix is needed, so a decision to run the old image is taken. And that's where the `flashwrite` functions take place.

**As suggested in the reviews and mentioned in the build output, I will divide this pull request into two seperate ones: one for fixing the bug in `suit` module, and one for adding the `flashwrite` functions**

File: sys/include/riotboot/flashwrite.h [hunk 10]
```diff
@@ -224,6 +229,30 @@ static inline int riotboot_flashwrite_finish(riotboot_flashwrite_t *state)
...
+ * @param[in]   slot       Target slot to invalidate
+ *
+ * @returns     0 on success, <0 otherwise
+ */
+int riotboot_flashwrite_invalidate(int slot);
+
+/**
+ * @brief   Revert a firmware update (riotboot version)
+ *
+ * This function clears the slot having the most recent firmware revision.
```
Comment (fjmolinas) [inline sys/include/riotboot/flashwrite.h:250]: I think we need to be a little more accurate here, it does not clear the slot but only invalidate it writing garbage to the header magic number.

Comment (fjmolinas) [reply sys/include/riotboot/flashwrite.h:250]: This also technically does not revert a firmware update, it just invalidates the running image A reboot is needed to "revert". I think `riotboot_flashwrite_invalidate_newest` or something like that would be more accurate.

File: sys/riotboot/flashwrite.c [hunk 11]
```diff
@@ -150,6 +151,33 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
...
+    }
+    if ((riotboot_slot_validate(riotboot_slot_other()) != 0) || (riotboot_slot_validate(riotboot_slot_current()) != 0)){
+        LOG_INFO(LOG_PREFIX "There will be no valid images to run after reboot\n");
+        return -2;
+    }
+
+    uint8_t data_flash[4];
+    memset(data_flash,INVALIDATE_HDR,sizeof(data_flash));
+
+    flashpage_write_raw((void *) riotboot_slot_get_hdr(slot), (const void *) data_flash, sizeof(RIOTBOOT_MAGIC));
```
Comment (benpicco) [inline sys/riotboot/flashwrite.c:168]:
```suggestion
    flashpage_write(riotboot_slot_get_hdr(slot), data_flash, sizeof(data_flash));
```

File: sys/riotboot/flashwrite.c [hunk 12]
```diff
@@ -150,6 +151,33 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
...
+        LOG_WARNING(LOG_PREFIX "Only one slot configured\n");
+        return -1;
+    }
+    if ((riotboot_slot_validate(riotboot_slot_other()) != 0) || (riotboot_slot_validate(riotboot_slot_current()) != 0)){
+        LOG_INFO(LOG_PREFIX "There will be no valid images to run after reboot\n");
+        return -2;
+    }
+
+    uint8_t data_flash[4];
+    memset(data_flash,INVALIDATE_HDR,sizeof(data_flash));
```
Comment (benpicco) [inline sys/riotboot/flashwrite.c:166]:
```suggestion
    memset(data_flash, INVALIDATE_HDR, sizeof(data_flash));
```

File: sys/riotboot/flashwrite.c [hunk 13]
```diff
@@ -150,6 +151,33 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
     return 0;
 }
 
+int riotboot_flashwrite_invalidate(int slot)
+{
+    if (riotboot_slot_numof == 1){
+        LOG_WARNING(LOG_PREFIX "Only one slot configured\n");
```
Comment (fjmolinas) [inline sys/riotboot/flashwrite.c:157]:
```suggestion
 if (riotboot_slot_numof == 1) {
        LOG_WARNING(LOG_PREFIX "abort, only one slot configured\n");
```

File: sys/riotboot/flashwrite.c [hunk 14]
```diff
@@ -150,6 +151,33 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
...
+        return -1;
+    }
+    if ((riotboot_slot_validate(riotboot_slot_other()) != 0) || (riotboot_slot_validate(riotboot_slot_current()) != 0)){
+        LOG_INFO(LOG_PREFIX "There will be no valid images to run after reboot\n");
+        return -2;
```
Comment (fjmolinas) [inline sys/riotboot/flashwrite.c:162]: This suggestion is up for discussion, the logic you have currently is requiring that both slots be valid, but IMO if you try to erase an already invalid slot, should this be allowed? I don't think this should return an error, so I adapted the logic that way.

```suggestion
    if (slot == riotboot_slot_current()) {
        if (!riotboot_slot_validate(riotboot_slot_other())) {
            LOG_WARNING(LOG_PREFIX "abort, slot %d is the only valid image\n", slot);
            return -2;
        }
    } else {
        if (!riotboot_slot_validate(riotboot_slot_current())) {
            LOG_WARNING(LOG_PREFIX "abort, slot %d is the only valid image\n", slot);
            return -2;
        }
```

**Review by fjmolinas — COMMENTED** [review]
@firas-hamdi I gave this one a test using #15559, I have some comments after testing. I also suggest a small change in the logs output.

let me know what you think about when `riotboot_flashwrite_invalidate` should return an error.

File: sys/riotboot/flashwrite.c (hunk 14 above)
Comment (fjmolinas) [reply sys/riotboot/flashwrite.c:162]: > I would not prefer to allow running the invalidation process twice on the same slot ( to invalidate the invalid).

Why? From a user perspective you would try to erase and get an error code when there is no reason for one. If you don't want to invalidate an already invalid slot than the user can implement that check IMO and avoid calling `riotboot_flashwrite_invalidate` altogether.

Comment (fjmolinas) [reply sys/riotboot/flashwrite.c:162]: > An action that I would suggest is to erase the flashpage of the invalid slot, only after reboot.

I don't why erasing the flashpage is related here.

Comment (firas-hamdi) [reply sys/riotboot/flashwrite.c:162]: > Why? From a user perspective you would try to erase and get an error code when there is no reason for one. If you don't want to invalidate an already invalid slot than the user can implement that check IMO and avoid calling `riotboot_flashwrite_invalidate` altogether.
>
I just wanted to avoid alike "NOP" state, that's the reason for the selected slot validity check. But as you said, I agree with the primary validity check from the user and thus not calling the `riotboot_flashwrite_invalidate` altogether.
I would just suggest this code for that logic, just to make the code shorter IMHO:
```suggestion
    if (riotboot_slot_validate (1 - slot) != 0) {
        LOG_WARNING(LOG_PREFIX "abort, slot %d is the only valid image\n", slot)
        return -2;
```
> 
> I don't why erasing the flashpage is related here.
>
No it was just a suggestion of a potential replacement to invalidate an invalid slot, I know it is related to another module.

Comment (fjmolinas) [reply sys/riotboot/flashwrite.c:162]: But in your suggestion you only validate one slot or am I missing something? Feel free to push your suggested change I can always comment on that after and its easier to view :)

Comment (firas-hamdi) [reply sys/riotboot/flashwrite.c]: Yes it's checking only the slot that has not been chosen to invalidate

File: sys/riotboot/flashwrite.c [hunk 15]
```diff
@@ -150,6 +151,33 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
...
+int riotboot_flashwrite_invalidate(int slot)
+{
+    if (riotboot_slot_numof == 1){
+        LOG_WARNING(LOG_PREFIX "abort, only one slot configured\n");
+        return -1;
+    }
+    if (riotboot_slot_validate(1 - slot) != 0){
+        LOG_WARNING(LOG_PREFIX "abort, slot %d is the only valid image\n", slot);
+        return -2;
+    }
```
Comment (fjmolinas) [inline sys/riotboot/flashwrite.c:163]: Both could be invalid though, so this statement is not necessarily accurate. Better phrase the other way around 
`abort, cant erase slot %d, other slot is not a valid, , slot`.

File: sys/include/riotboot/flashwrite.h [hunk 16]
```diff
@@ -224,6 +229,30 @@ static inline int riotboot_flashwrite_finish(riotboot_flashwrite_t *state)
...
+ * @param[in]   slot       Target slot to invalidate
+ *
+ * @returns     0 on success, <0 otherwise
+ */
+int riotboot_flashwrite_invalidate(int slot);
+
+/**
+ * @brief   Invalidate the latest firmware version (riotboot version)
+ *
+ * This function invalidates the slot having the most recent firmware revision.
```
Comment (fjmolinas) [inline sys/include/riotboot/flashwrite.h:250]: Should add a mention that it requires at least two valid images.

File: sys/riotboot/flashwrite.c [hunk 17]
```diff
@@ -150,6 +151,33 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
...
+    flashpage_write((void *)riotboot_slot_get_hdr(slot), data_flash, sizeof(data_flash));
+
+    return 0;
+}
+
+int riotboot_flashwrite_invalidate_latest(void)
+{
+    int _slot_to_revert;
+    _slot_to_revert = (riotboot_slot_get_hdr(riotboot_slot_other())->version
+            > riotboot_slot_get_hdr(riotboot_slot_current())->version) ? riotboot_slot_other() : riotboot_slot_current();
```
Comment (fjmolinas) [inline sys/riotboot/flashwrite.c:177]:
```suggestion
    _slot_to_revert = (riotboot_slot_get_hdr(riotboot_slot_other())->version
                       > riotboot_slot_get_hdr(riotboot_slot_current())->version)
                      ? riotboot_slot_other() : riotboot_slot_current();
```

File: sys/riotboot/flashwrite.c [hunk 18]
```diff
@@ -150,6 +151,33 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
...
 }
 
+int riotboot_flashwrite_invalidate(int slot)
+{
+    if (riotboot_slot_numof == 1){
+        LOG_WARNING(LOG_PREFIX "abort, only one slot configured\n");
+        return -1;
+    }
+    if (riotboot_slot_validate(1 - slot) != 0){
+        LOG_WARNING(LOG_PREFIX "abort, can not erase slot[%d], other slot[%d] is invalid\n", slot, 1 - slot);
```
Comment (fjmolinas) [inline sys/riotboot/flashwrite.c:161]:
```suggestion
        LOG_WARNING(LOG_PREFIX "abort, can not erase slot[%d], other slot[%d] is invalid\n",
            slot, 1 - slot);
```

File: sys/include/riotboot/flashwrite.h [hunk 19]
```diff
@@ -224,6 +229,32 @@ static inline int riotboot_flashwrite_finish(riotboot_flashwrite_t *state)
...
 }
 
+/**
+ * @brief   Invalidate a slot header (riotboot version)
+ *
+ * This function invalidates the target slot.
+ *
+ * @note    If this function is called with only one valid slot,
+ *          The invalidation will fail in order to keep one image
+ *          to run after reboot
```
Comment (fjmolinas) [inline sys/include/riotboot/flashwrite.h:239]:
```suggestion
 * @note    If this function is called with only one valid slot,
 *          the invalidation will fail in order to keep one valid
 *          image to run after reboot
```

**Review by fjmolinas — COMMENTED** [review]
Some minor nitpicks, please squash once addressed.

**Comment by fjmolinas** [comment]
@firas-hamdi please squash all into one commit.

File: sys/include/riotboot/flashwrite.h [hunk 20]
```diff
@@ -128,6 +128,11 @@ typedef struct {
  */
 #define RIOTBOOT_FLASHWRITE_SKIPLEN     sizeof(RIOTBOOT_MAGIC)
 
+/**
+ * @brief Magic number used to invalidate a slot
+ */
+#define INVALIDATE_HDR                  0xAA
+
```
Comment (benpicco) [inline sys/include/riotboot/flashwrite.h:135]: No need to have this in the global namespace - just move it to `sys/riotboot/flashwrite.c`

File: sys/riotboot/flashwrite.c [hunk 21]
```diff
@@ -150,6 +151,34 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
     return 0;
 }
 
+int riotboot_flashwrite_invalidate(int slot)
+{
+    if (riotboot_slot_numof == 1){
```
Comment (benpicco) [inline sys/riotboot/flashwrite.c:156]:
```suggestion
    if (riotboot_slot_numof == 1) {
```

File: sys/riotboot/flashwrite.c [hunk 22]
```diff
@@ -150,6 +151,34 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
...
+        LOG_WARNING(LOG_PREFIX "abort, only one slot configured\n");
+        return -1;
+    }
+    if (riotboot_slot_validate(1 - slot) != 0){
```
Comment (benpicco) [inline sys/riotboot/flashwrite.c:160]:
```suggestion
    if (riotboot_slot_validate(1 - slot) != 0) {
```

**Review by benpicco — COMMENTED** [review]
Looks good to me, just some minor nitpicks - you can squash directly.

**Review by benpicco — APPROVED** [review]
looks good to me
